src/Interpreters/Cluster.cpp

```diff
@@ -573,6 +574,8 @@ Cluster::Cluster(Cluster::ReplicasAsShardsTag, const Cluster & from, const Setti
                 continue;   /// Duplicate host, skip.
 
             ShardInfo info;
+            info.shard_num = ++shard_num;
```

Comment:
I think it's correct. This function only used for `clusterAllReplicas` which regard all replicas as independent shards.

Reply:
It create a new `Cluster` contains all replicas as shard.

Reply:
Let's just add functional test that shows that everything is correct for multi-replica shards.

Reply:
How can I add another cluster for test?

Reply:
@ucasFL You can edit `tests/config.d/cluster.xml`

Reply:
Why I add cluster definition into server-test.xml or cluster.xml, the test still also show that cluster not found? @alexey-milovidov

Reply:
>Why I add cluster definition into server-test.xml or cluster.xml, the test still also show that cluster not found?

AFAIK server-test.xml is not used anywhere (except for arcadia maybe), if you will add your cluster to `tests/config/config.d/clusters.xml` it should work.

FWIW fast-test (and other checks) got this configs via tests/config/install.sh, it already includes `clusters.xml`, here is a snippet from fast-test log:
```
+ ln -sf /ClickHouse/tests/config/config.d/clusters.xml /fasttest-workspace/db-fasttest/config.d/
```

Previous variant (https://github.com/ClickHouse/ClickHouse/commit/f6e83b579cc2b9db37b0a6315ee5633eaa6cf86a) - https://clickhouse-test-reports.s3.yandex.net/21498/f6e83b579cc2b9db37b0a6315ee5633eaa6cf86a/fast_test.html#fail1 was failed because you set port to 59000, and this port is used only for arcadia tests AFAIR.

So to summarize you should add everything like in f6e83b579cc2b9db37b0a6315ee5633eaa6cf86a, but with default port.